### ISSUE TITLE
Program killed by SIGFPE in the backward pass of torch.sparse.sampled_addmm when input has specific shape

### ISSUE THREAD

**Issue by Elysion-tcfa** [pytorch/pytorch#117431]
### 🐛 Describe the bug

Recently I'm trying to implement graph attention using PyTorch, so I'm utilizing `torch.sparse.sampled_addmm`. Unfortunately, I find that the backward pass crashes if the sparse `input` has specific shape, but not necessarily a lot of nonzero elements. Here is a simple script to reproduce:
```
N = 65536
mask = torch.sparse_coo_tensor(torch.stack([torch.arange(N), torch.arange(N)], dim=0), torch.ones(N)).cuda().to_sparse_csr()
mask.requires_grad = True
q = torch.randn(N, 128).cuda()
q.requires_grad = True
k = torch.randn(N, 128).cuda()
k.requires_grad = True
attn = torch.sparse.sampled_addmm(mask, q, k.t(), beta=0)
attn.backward(torch.ones_like(attn))
```
This script evaluates `q @ k.t()` only at the diagonal. Unfortunately, this script crashes under SIGFPE error signal. It seems that this error only happens when `N` is a multiple of 65536. The error is gone when `N` is 32768, 65537 or 98304 (32768\*3), but persists when `N` is changed to 131072 or 196608 (65536\*3). It also doesn't seem to depend on the number of nonzero elements. For example, if `mask` is changed to block diagonal with blocks of ones, the behavior is the same:
```
indices = torch.repeat_interleave(torch.arange(N), 128)
indices = torch.stack([indices, torch.arange(len(indices)) % 128 + indices // 128 * 128], dim=0)
mask = torch.sparse_coo_tensor(indices, torch.ones(indices.shape[1])).cuda().to_sparse_csr()
```

In the meantime, the workaround is to create a `torch.autograd.Function` which wraps `torch.sparse.sampled_addmm` and manually implement the gradients using `torch.mm`.
```
class SampledMMFunction(torch.autograd.Function):
    @staticmethod
    def forward(ctx, input, mat1, mat2):
        ctx.save_for_backward(mat1, mat2)
        return torch.sparse.sampled_addmm(input, mat1, mat2, beta=0)

    @staticmethod
    def backward(ctx, grad_output):
        mat1, mat2 = ctx.saved_tensors
        return torch.zeros_like(grad_output), torch.mm(grad_output, mat2.t()), torch.mm(mat1.t(), grad_output)
```

### Versions

PyTorch version: 2.1.1
Is debug build: False
CUDA used to build PyTorch: 12.3
ROCM used to build PyTorch: N/A

OS: Arch Linux (x86_64)
GCC version: (GCC) 13.2.1 20230801
Clang version: Could not collect
CMake version: version 3.27.8
Libc version: glibc-2.38

Python version: 3.11.6 (main, Nov 14 2023, 09:36:21) [GCC 13.2.1 20230801] (64-bit runtime)
Python platform: Linux-6.6.2-arch1-1-x86_64-with-glibc2.38
Is CUDA available: True
CUDA runtime version: 12.3.52
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4060 Laptop GPU
Nvidia driver version: 545.29.06
cuDNN version: Probably one of the following:
/usr/lib/libcudnn.so.8.9.6
/usr/lib/libcudnn_adv_infer.so.8.9.6
/usr/lib/libcudnn_adv_train.so.8.9.6
/usr/lib/libcudnn_cnn_infer.so.8.9.6
/usr/lib/libcudnn_cnn_train.so.8.9.6
/usr/lib/libcudnn_ops_infer.so.8.9.6
/usr/lib/libcudnn_ops_train.so.8.9.6
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      48 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             16
On-line CPU(s) list:                0-15
Vendor ID:                          AuthenticAMD
Model name:                         AMD Ryzen 9 7940HS w/ Radeon 780M Graphics
CPU family:                         25
Model:                              116
Thread(s) per core:                 2
Core(s) per socket:                 8
Socket(s):                          1
Stepping:                           1
CPU(s) scaling MHz:                 32%
CPU max MHz:                        6228.0000
CPU min MHz:                        400.0000
BogoMIPS:                           7988.30
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good amd_lbr_v2 nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba perfmon_v2 ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local user_shstk avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold v_vmsave_vmload vgif x2avic v_spec_ctrl vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid overflow_recov succor smca flush_l1d
Virtualization:                     AMD-V
L1d cache:                          256 KiB (8 instances)
L1i cache:                          256 KiB (8 instances)
L2 cache:                           8 MiB (8 instances)
L3 cache:                           16 MiB (1 instance)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-15
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Vulnerable: Safe RET, no microcode
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced / Automatic IBRS, IBPB conditional, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.2
[pip3] torch==2.1.1
[conda] Could not collect
